libs/icu_headroom_metric.py

```diff
@@ -0,0 +1,168 @@
+from typing import Union, Optional
+import enum
+from covidactnow.datapublic.common_fields import CommonFields
+from libs.datasets import can_model_output_schema as schema
+
+
+import pandas as pd
+
+
+class CovidPatientsMethod(enum.Enum):
+    ACTUAL = 0
+    ESTIMATED = 1
+
+
+class NonCovidPatientsMethod(enum.Enum):
+    ACTUAL = 0
+    ESTIMATED_FROM_TYPICAL_UTILIZATION = 1
+    ESTIMATED_FROM_TOTAL_ICU_ACTUAL = 2
+
+
+DEFAULT_ICU_DECOMP = 0.21
+
+ICU_DECOMP_OVERRIDE = {
+    "AL": 0.15,
+    "AZ": 0.4,
+    "DE": 0.3,
+    "DC": 0.15,
+    "GA": 0.1,
+    # TODO(https://trello.com/c/1ddB5ntl/): CCM is currently giving us an
+    # extra-high utilization rate. If that gets fixed we may need to bump this
+    # back down.
+    "MS": 0.37,
+    "NV": 0.25,
+    "RI": 0,
+}
+
+
+def get_decomp_for_state(state: str) -> float:
+    return ICU_DECOMP_OVERRIDE.get(state, DEFAULT_ICU_DECOMP)
+
+
+class ICUMetricData:
+    def __init__(self, data, latest_values, decomp_factor):
+        self.data = data
+        self.latest_values = latest_values
+        self.decomp_factor = decomp_factor
+
+    @property
+    def actual_current_icu_covid(self) -> Optional[pd.Series]:
+        actuals = self.data[CommonFields.CURRENT_ICU]
+        if not actuals.any():
+            return None
+
+        return actuals
+
+    @property
+    def estimated_current_icu_covid(self) -> Optional[pd.Series]:
+        estimated = self.data[schema.CURRENT_ICU]
+        if not estimated.any():
+            return None
+
+        return estimated
+
+    @property
+    def actual_current_icu_total(self) -> Optional[pd.Series]:
+        actuals = self.data[CommonFields.CURRENT_ICU_TOTAL]
+        if not actuals.any():
+            return None
+
+        return actuals
+
+    @property
+    def total_icu_beds(self):
```

Comment:
nit: Is `self.latest_values[CommonFields.ICU_BEDS]` a scalar value (as opposed to timeseries)?  So this may return a pd.Series or a scalar value?  Could you document that in the return type like you do for the above property definitions in this file?

In general, I'm curious how you're deciding when to document the return type...  I'm a fan of always documenting types but I have no idea what is idiomatic in python. 😁

Reply:
ah I'm generally very very pro return types in python! I just hadn't gone through and cleaned up the type annotations here. I do try to have return types as much as possible, but sometimes add them as a last step